node/node_cross_shard.go

```diff
@@ -64,6 +64,33 @@ func (node *Node) BroadcastCXReceipts(newBlock *types.Block, lastCommits []byte)
 	}
 }
 
+// BroadcastMissingCXReceipts broadcasts missing cross shard receipts per request
+func (node *Node) BroadcastMissingCXReceipts() {
+	sendNextTime := []common.Hash{}
+	it := node.CxPool.Pool().Iterator()
+	for blockHash := range it.C {
+		blk := node.Blockchain().GetBlockByHash(blockHash.(common.Hash))
+		if blk == nil {
+			continue
+		}
+		blockNum := blk.NumberU64()
+		nextBlk := node.Blockchain().GetBlockByNumber(blockNum + 1)
+		if nextBlk == nil {
+			sendNextTime = append(sendNextTime, blockHash.(common.Hash))
+			continue
+		}
+		sig := nextBlk.Header().LastCommitSignature()
+		bitmap := nextBlk.Header().LastCommitBitmap()
+		lastCommits := append(sig[:], bitmap...)
+		go node.BroadcastCXReceipts(blk, lastCommits)
```

Comment:
Doesn't this broadcast to all N shards?  Can we make it broadcast to just the shard in which the tx is found?

(In order for this to work, you might have to track `(blockHash, toShard)` tuple in the pool.)

Reply:
changed the data structure to be (blockHash, toShardID); also added tx type checking